### PR TITLE
Enabling Sort Optimization to make use of Lucene

### LINKED ISSUE #1570: search_after queries are not performing as expected
#### Disclaimer: the origin of this issue was a [post](https://discuss.elastic.co/t/search-after-queries-are-not-performing-as-expected/289174) in ElasticSearch discuss.

**Is your feature request related to a problem? Please describe.**

We began to use the search_after parameter for solving our issues related to pagination consistency. This works fine until we realized that our query slow log started to have a lot of entries. At some moments this lead us to high CPU usage and high search latency, which degraded our service.

All of these log entries show us queries without filters or search terms and use the search_after parameter. This kind of query is taking ~400ms to be resolved. The issue is that its equivalent using a simple from/size pagination is taking just ~20ms.

So, why ES is taking more time/resources to resolve the query with "search_after" than its equivalent with from/size?

**Describe the solution you'd like**
After some query profiling, the team of ElasticSearch came with the explanation:

> I see what's happening.
> 
> From 7.6 we introduced an optimization 2 for sort queries, we rewrite a sort query as LongDistanceFeatureQuery that can skip over blocks of documents. This makes sort queries under certain conditions much faster. The optimization is currently not applied to search_after case. From 7.16 it will be applied to search_after as well, so sort queries with search_after also will be much faster under certain conditions.

So, this issue is just to ask for including this optimization to improve search_after queries in OpenSearch too.

**Describe alternatives you've considered**
In the meantime, the only thing we can do is add index sorting, which improves dramatically the issue, but just for one sort criteria.

**Additional context**

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
This PR removes sort optimization code present in OpenSearch because from Lucence 8.10.1, Numeric sort optimization is already done in Lucene Comparators. Now Lucene also supports sort optimization with search_after queries.

 
### Issues Resolved
#1570 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Rishikesh1159** [comment]
Below are OpenSearch Benchmark results comparing current master branch with this PR branch:
 We will be looking at tasks which might be effected by this change and have major service time difference.

 Regular Sort :

->  asc_sort_population is 15x faster.
-> desc_sort_population is 20x faster.
-> asc_sort_geonameid does not significant improvement	
-> desc_sort_geonameid has slight improvement


Sort with after :

-> asc_sort_with_after_population is 20x faster.
-> asc_sort_with_after_geonameid is 10x faster.
-> desc_sort_with _after_geonameid is 3x faster.



Current Master Branch:

|                                                         Metric |                           Task |       Value |    Unit |
|---------------------------------------------------------------:|-------------------------------:|------------:|--------:|
|                                        50th percentile latency |           desc_sort_population |     100.415 |      ms |
|                                        90th percentile latency |           desc_sort_population |      175.17 |      ms |
|                                        99th percentile latency |           desc_sort_population |      189.38 |      ms |
|                                       100th percentile latency |           desc_sort_population |     189.699 |      ms |
|                                   50th percentile service time |           desc_sort_population |     98.1873 |      ms |
|                                   90th percentile service time |           desc_sort_population |     173.868 |      ms |
|                                   99th percentile service time |           desc_sort_population |     187.129 |      ms |
|                                  100th percentile service time |           desc_sort_population |     188.619 |      ms |
|                                        50th percentile latency |            asc_sort_population |     99.3038 |      ms |
|                                        90th percentile latency |            asc_sort_population |     112.862 |      ms |
|                                        99th percentile latency |            asc_sort_population |     185.557 |      ms |
|                                       100th percentile latency |            asc_sort_population |     186.135 |      ms |
|                                   50th percentile service time |            asc_sort_population |     97.1404 |      ms |
|                                   90th percentile service time |            asc_sort_population |     110.553 |      ms |
|                                   99th percentile service time |            asc_sort_population |     183.189 |      ms |
|                                  100th percentile service time |            asc_sort_population |     184.295 |      ms |
|                                        50th percentile latency | asc_sort_with_after_population |     144.535 |      ms |
|                                        90th percentile latency | asc_sort_with_after_population |     220.117 |      ms |
|                                        99th percentile latency | asc_sort_with_after_population |     250.599 |      ms |
|                                       100th percentile latency | asc_sort_with_after_population |     262.809 |      ms |
|                                   50th percentile service time | asc_sort_with_after_population |     142.943 |      ms |
|                                   90th percentile service time | asc_sort_with_after_population |      218.88 |      ms |
|                                   99th percentile service time | asc_sort_with_after_population |     249.452 |      ms |
|                                  100th percentile service time | asc_sort_with_after_population |     261.684 |      ms |
|                                        50th percentile latency |            desc_sort_geonameid |     9.36181 |      ms |
|                                        90th percentile latency |            desc_sort_geonameid |     13.1581 |      ms |
|                                        99th percentile latency |            desc_sort_geonameid |      18.971 |      ms |
|                                       100th percentile latency |            desc_sort_geonameid |     23.2126 |      ms |
|                                   50th percentile service time |            desc_sort_geonameid |     8.19783 |      ms |
|                                   90th percentile service time |            desc_sort_geonameid |     12.2265 |      ms |
|                                   99th percentile service time |            desc_sort_geonameid |     16.6765 |      ms |
|                                  100th percentile service time |            desc_sort_geonameid |     22.3743 |      ms |
|                                        50th percentile latency | desc_sort_with_after_geonameid |     123.608 |      ms |
|                                        90th percentile latency | desc_sort_with_after_geonameid |     207.159 |      ms |
|                                        99th percentile latency | desc_sort_with_after_geonameid |     233.925 |      ms |
|                                       100th percentile latency | desc_sort_with_after_geonameid |     248.288 |      ms |
|                                   50th percentile service time | desc_sort_with_after_geonameid |     120.187 |      ms |
|                                   90th percentile service time | desc_sort_with_after_geonameid |     206.102 |      ms |
|                                   99th percentile service time | desc_sort_with_after_geonameid |     232.728 |      ms |
|                                  100th percentile service time | desc_sort_with_after_geonameid |      242.52 |      ms |
|                                        50th percentile latency |             asc_sort_geonameid |     8.16044 |      ms |
|                                        90th percentile latency |             asc_sort_geonameid |     10.2532 |      ms |
|                                        99th percentile latency |             asc_sort_geonameid |      11.268 |      ms |
|                                       100th percentile latency |             asc_sort_geonameid |     11.4913 |      ms |
|                                   50th percentile service time |             asc_sort_geonameid |     6.93547 |      ms |
|                                   90th percentile service time |             asc_sort_geonameid |     9.16224 |      ms |
|                                   99th percentile service time |             asc_sort_geonameid |     9.73752 |      ms |
|                                  100th percentile service time |             asc_sort_geonameid |     9.94709 |      ms |
|                                        50th percentile latency |  asc_sort_with_after_geonameid |      111.78 |      ms |
|                                        90th percentile latency |  asc_sort_with_after_geonameid |     187.119 |      ms |
|                                        99th percentile latency |  asc_sort_with_after_geonameid |     215.821 |      ms |
|                                       100th percentile latency |  asc_sort_with_after_geonameid |     223.899 |      ms |
|                                   50th percentile service time |  asc_sort_with_after_geonameid |     110.322 |      ms |
|                                   90th percentile service time |  asc_sort_with_after_geonameid |     186.074 |      ms |
|                                   99th percentile service time |  asc_sort_with_after_geonameid |     215.349 |      ms |
|                                  100th percentile service time |  asc_sort_with_after_geonameid |     222.589 |      ms |


PR branch :

|                                                         Metric |                           Task |       Value |    Unit |
|---------------------------------------------------------------:|-------------------------------:|------------:|--------:|
|                                        50th percentile latency |           desc_sort_population |     7.01302 |      ms |
|                                        90th percentile latency |           desc_sort_population |     7.96208 |      ms |
|                                        99th percentile latency |           desc_sort_population |     9.22419 |      ms |
|                                       100th percentile latency |           desc_sort_population |     10.2252 |      ms |
|                                   50th percentile service time |           desc_sort_population |     5.20505 |      ms |
|                                   90th percentile service time |           desc_sort_population |     6.23516 |      ms |
|                                   99th percentile service time |           desc_sort_population |     7.52332 |      ms |
|                                  100th percentile service time |           desc_sort_population |     7.97527 |      ms |
|                                        50th percentile latency |            asc_sort_population |     3.99154 |      ms |
|                                        90th percentile latency |            asc_sort_population |     4.59955 |      ms |
|                                        99th percentile latency |            asc_sort_population |     5.31723 |      ms |
|                                       100th percentile latency |            asc_sort_population |     13.7694 |      ms |
|                                   50th percentile service time |            asc_sort_population |     2.38935 |      ms |
|                                   90th percentile service time |            asc_sort_population |     2.76126 |      ms |
|                                   99th percentile service time |            asc_sort_population |     3.19022 |      ms |
|                                  100th percentile service time |            asc_sort_population |     11.7281 |      ms |
|                                        50th percentile latency | asc_sort_with_after_population |     9.52427 |      ms |
|                                        90th percentile latency | asc_sort_with_after_population |     13.3029 |      ms |
|                                        99th percentile latency | asc_sort_with_after_population |     14.2207 |      ms |
|                                       100th percentile latency | asc_sort_with_after_population |      14.475 |      ms |
|                                   50th percentile service time | asc_sort_with_after_population |      7.8274 |      ms |
|                                   90th percentile service time | asc_sort_with_after_population |     11.5684 |      ms |
|                                   99th percentile service time | asc_sort_with_after_population |     12.2756 |      ms |
|                                  100th percentile service time | asc_sort_with_after_population |     12.4841 |      ms |
|                                        50th percentile latency |            desc_sort_geonameid |     9.88083 |      ms |
|                                        90th percentile latency |            desc_sort_geonameid |     11.5916 |      ms |
|                                        99th percentile latency |            desc_sort_geonameid |     12.0913 |      ms |
|                                       100th percentile latency |            desc_sort_geonameid |     12.5845 |      ms |
|                                   50th percentile service time |            desc_sort_geonameid |     8.50975 |      ms |
|                                   90th percentile service time |            desc_sort_geonameid |     10.3373 |      ms |
|                                   99th percentile service time |            desc_sort_geonameid |     11.1854 |      ms |
|                                  100th percentile service time |            desc_sort_geonameid |     11.2994 |      ms |
|                                        50th percentile latency | desc_sort_with_after_geonameid |     48.6053 |      ms |
|                                        90th percentile latency | desc_sort_with_after_geonameid |     56.5896 |      ms |
|                                        99th percentile latency | desc_sort_with_after_geonameid |     57.3231 |      ms |
|                                       100th percentile latency | desc_sort_with_after_geonameid |     57.3962 |      ms |
|                                   50th percentile service time | desc_sort_with_after_geonameid |     47.1192 |      ms |
|                                   90th percentile service time | desc_sort_with_after_geonameid |     55.2823 |      ms |
|                                   99th percentile service time | desc_sort_with_after_geonameid |     55.5241 |      ms |
|                                  100th percentile service time | desc_sort_with_after_geonameid |     55.7547 |      ms |
|                                        50th percentile latency |             asc_sort_geonameid |     8.39487 |      ms |
|                                        90th percentile latency |             asc_sort_geonameid |     8.93019 |      ms |
|                                        99th percentile latency |             asc_sort_geonameid |     9.45887 |      ms |
|                                       100th percentile latency |             asc_sort_geonameid |     10.3578 |      ms |
|                                   50th percentile service time |             asc_sort_geonameid |     7.21012 |      ms |
|                                   90th percentile service time |             asc_sort_geonameid |     7.51989 |      ms |
|                                   99th percentile service time |             asc_sort_geonameid |     8.30312 |      ms |
|                                  100th percentile service time |             asc_sort_geonameid |     9.45062 |      ms |
|                                        50th percentile latency |  asc_sort_with_after_geonameid |      10.942 |      ms |
|                                        90th percentile latency |  asc_sort_with_after_geonameid |     13.6304 |      ms |
|                                        99th percentile latency |  asc_sort_with_after_geonameid |     15.5996 |      ms |
|                                       100th percentile latency |  asc_sort_with_after_geonameid |     17.1587 |      ms |
|                                   50th percentile service time |  asc_sort_with_after_geonameid |     9.58448 |      ms |
|                                   90th percentile service time |  asc_sort_with_after_geonameid |     12.2301 |      ms |
|                                   99th percentile service time |  asc_sort_with_after_geonameid |      14.795 |      ms |
|                                  100th percentile service time |  asc_sort_with_after_geonameid |     16.4617 |      ms |

File: server/src/main/java/org/opensearch/search/query/QueryPhase.java [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (nknize) [inline server/src/main/java/org/opensearch/search/query/QueryPhase.java:1]: accidental extra line insertion

File: server/src/test/java/org/opensearch/search/query/QueryPhaseTests.java [hunk 2]
```diff
@@ -114,13 +106,8 @@
 import java.util.Collections;
 import java.util.List;
 
-import static org.opensearch.search.query.QueryPhase.pointsHaveDuplicateData;
+import static org.hamcrest.Matchers.*;
```
Comment (nknize) [inline server/src/test/java/org/opensearch/search/query/QueryPhaseTests.java:109]: oopsie wildcard insert

**Review by nknize — CHANGES_REQUESTED** [review]
:tada:  thx for taking care of this! First pass just a few nits.

**Comment by nknize** [comment]
> Below are OpenSearch Benchmark results

:tada:  Nice! Thx for including the benchmarks!! Could we re-format to the following joined table to make it easier to compare?

|                                                         Metric |                           Task |       Value (Baseline) |    Value (PR)  |    Unit |
|---------------------------------------------------------------:|-------------------------------:|------------:|--------:|------------:|

**Comment by Rishikesh1159** [comment]
> > Below are OpenSearch Benchmark results
> 
> 🎉 Nice! Thx for including the benchmarks!! Could we re-format to the following joined table to make it easier to compare?
> 
> Metric	Task	Value (Baseline)	Value (PR)	Unit

@nknize Sure I will re-format and update the comment

File: server/src/test/java/org/opensearch/search/query/QueryPhaseTests.java (hunk 2 above)
Comment (Rishikesh1159) [reply server/src/test/java/org/opensearch/search/query/QueryPhaseTests.java:109]: Thank you for catching it, I will fix this

**Review by nknize — APPROVED** [review]
LGTM. Thx!

**Comment by nknize** [comment]
Could you please open a backport PR as well?

**Comment by Rishikesh1159** [comment]
@nknize Sure I will open a backport PR. Thanks for your review!

**Pull request merged**